### PR TITLE
Don't force the file extension to .js or .wasm when compiling side modules

### PR DESCRIPTION
**Pull request by dos1**
Specifying `-s SIDE_MODULE=1` (or `=2`) is already enough to determine the file type. Emscripten's preload plugins already expect WASM side modules to have `.so` extension, so this fixes a peculiar situation where Emscripten can't produce a file in a way Emscripten itself expects it to be later.

PS. I really hate this `and 'SIDE_MODULE=1' not in settings_changes and 'SIDE_MODULE=2' not in settings_changes`, but I didn't know how to do it cleanly without reorganizing everything around...

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -949,7 +949,7 @@ def get_last_setting_change(setting):
 
     # If not compiling to JS, then we are compiling to an intermediate bitcode objects or library, so
     # ignore dynamic linking, since multiple dynamic linkings can interfere with each other
-    if filename_type_suffix(target) not in JS_CONTAINING_SUFFIXES or options.ignore_dynamic_linking:
+    if (filename_type_suffix(target) not in JS_CONTAINING_SUFFIXES and 'SIDE_MODULE=1' not in settings_changes and 'SIDE_MODULE=2' not in settings_changes) or options.ignore_dynamic_linking:
```
Comment (kripken) [inline emcc.py]: can use `settings_key_changes` which is a set of all user-changed settings

File: emcc.py [hunk 2]
```diff
@@ -1054,6 +1054,11 @@ def check(input_file):
     elif shared.Settings.SIDE_MODULE:
       assert not shared.Settings.MAIN_MODULE
       options.memory_init_file = False # memory init file is not supported with asm.js side modules, must be executable synchronously (for dlopen)
+      # SIDE_MODULE+WASM combo unambiguously specifies what kind of file we want to output, so we don't have to fiddle with file extension
+      if shared.Settings.WASM:
+        wasm_binary_target = target
+      else:
+        js_target = target
```
Comment (kripken) [inline emcc.py]: why is this line needed, the js path should still work as before?

File: emcc.py [hunk 3]
```diff
@@ -1564,7 +1569,7 @@ def compile_source_file(i, input_file):
               temp_files[pos] = (temp_files[pos][0], new_temp_file)
 
       # Decide what we will link
-      stop_at_bitcode = final_suffix not in EXECUTABLE_SUFFIXES
+      stop_at_bitcode = final_suffix not in EXECUTABLE_SUFFIXES and not shared.Settings.SIDE_MODULE
```
Comment (kripken) [inline emcc.py]: this line looks surprising. it seems like we should add a method that checks the suffix, and call it from here, and in there have a comment about this.

**Comment by kripken** [comment]
(that is, add `.so` to either `WASM_ENDINGS` or `EXECUTABLE_SUFFIXES`)

**Comment by dos1** [comment]
As far as I understand it, adding `.so` to the list would still require some checks on whether `SIDE_MODULE` is enabled or not, because otherwise it would break building to bytecode with `.so` extension (which is something many build systems rely on) - so I opted for a simpler approach to just allow everything when `SIDE_MODULE` is specified.

File: emcc.py (hunk 2 above)
Comment (dos1) [reply emcc.py]: For consistency, as one might want to always build their side modules to `.so` files, regardless of the build type (wasm or asmjs), to make the filename to `dlopen` predictable. Not really attached to that though.

**Comment by kripken** [comment]
It's true that if all we allow is `.so` when SIDE_MODULE then we'd need to check that. But it seems simple, about the same.

Do you see a benefit to other suffixes? Otherwise, I think just adding `.so` is best. In particular then we'd match the `.so` expectation of the dynamic loader code, as you pointed out.

**Comment by kripken** [comment]
I think this was closed by mistake when we deleted the incoming branch (after switching to master). If this PR is still relevant please reopen.

**Comment by sbc100** [comment]
The deletion of the `incoming` branch triggered the automatic closing of this branch.  My apologies.  This was an unexpected side effect.

Re-opening and re-targeting to master.
